README
NAME
    POE::Component::Client::HTTP - a HTTP user-agent component

VERSION
    version 0.948

SYNOPSIS
      use POE qw(Component::Client::HTTP);

      POE::Component::Client::HTTP->spawn(
        Agent     => 'SpiffCrawler/0.90',   # defaults to something long
        Alias     => 'ua',                  # defaults to 'weeble'
        From      => 'spiffster@perl.org',  # defaults to undef (no header)
        Protocol  => 'HTTP/0.9',            # defaults to 'HTTP/1.1'
        Timeout   => 60,                    # defaults to 180 seconds
        MaxSize   => 16384,                 # defaults to entire response
        Streaming => 4096,                  # defaults to 0 (off)
        FollowRedirects => 2,               # defaults to 0 (off)
        Proxy     => "http://localhost:80", # defaults to HTTP_PROXY env. variable
        NoProxy   => [ "localhost", "127.0.0.1" ], # defs to NO_PROXY env. variable
        BindAddr  => "12.34.56.78",         # defaults to INADDR_ANY
      );

      $kernel->post(
        'ua',        # posts to the 'ua' alias
        'request',   # posts to ua's 'request' state
        'response',  # which of our states will receive the response
        $request,    # an HTTP::Request object
      );

      # This is the sub which is called when the session receives a
      # 'response' event.
      sub response_handler {
        my ($request_packet, $response_packet) = @_[ARG0, ARG1];

        # HTTP::Request
        my $request_object  = $request_packet->[0];

        # HTTP::Response
        my $response_object = $response_packet->[0];

        my $stream_chunk;
        if (! defined($response_object->content)) {
          $stream_chunk = $response_packet->[1];
        }

        print(
          "*" x 78, "\n",
          "*** my request:\n",
          "-" x 78, "\n",
          $request_object->as_string(),
          "*" x 78, "\n",
          "*** their response:\n",
          "-" x 78, "\n",
          $response_object->as_string(),
        );

        if (defined $stream_chunk) {
          print "-" x 40, "\n", $stream_chunk, "\n";
        }

        print "*" x 78, "\n";
      }

DESCRIPTION
    POE::Component::Client::HTTP is an HTTP user-agent for POE. It lets
    other sessions run while HTTP transactions are being processed, and it
    lets several HTTP transactions be processed in parallel.

    It supports keep-alive through POE::Component::Client::Keepalive, which
    in turn uses POE::Component::Resolver for asynchronous IPv4 and IPv6
    name resolution.

    HTTP client components are not proper objects. Instead of being created,
    as most objects are, they are "spawned" as separate sessions. To avoid
    confusion (and hopefully not cause other confusion), they must be
    spawned with a "spawn" method, not created anew with a "new" one.

CONSTRUCTOR
  spawn
    PoCo::Client::HTTP's "spawn" method takes a few named parameters:

    Agent => $user_agent_string
    Agent => \@list_of_agents
      If a UserAgent header is not present in the HTTP::Request, a random
      one will be used from those specified by the "Agent" parameter. If
      none are supplied, POE::Component::Client::HTTP will advertise itself
      to the server.

      "Agent" may contain a reference to a list of user agents. If this is
      the case, PoCo::Client::HTTP will choose one of them at random for
      each request.

    Alias => $session_alias
      "Alias" sets the name by which the session will be known. If no alias
      is given, the component defaults to "weeble". The alias lets several
      sessions interact with HTTP components without keeping (or even
      knowing) hard references to them. It's possible to spawn several HTTP
      components with different names.

    ConnectionManager => $poco_client_keepalive
      "ConnectionManager" sets this component's connection pool manager. It
      expects the connection manager to be a reference to a
      POE::Component::Client::Keepalive object. The HTTP client component
      will call "allocate()" on the connection manager itself so you should
      not have done this already.

        my $pool = POE::Component::Client::Keepalive->new(
          keep_alive    => 10, # seconds to keep connections alive
          max_open      => 100, # max concurrent connections - total
          max_per_host  => 20, # max concurrent connections - per host
          timeout       => 30, # max time (seconds) to establish a new connection
        );

        POE::Component::Client::HTTP->spawn(
          # ...
          ConnectionManager => $pool,
          # ...
        );

      See POE::Component::Client::Keepalive for more information, including
      how to alter the connection manager's resolver configuration (for
      example, to force IPv6 or prefer it before IPv4).

    CookieJar => $cookie_jar
      "CookieJar" sets the component's cookie jar. It expects the cookie jar
      to be a reference to a HTTP::Cookies object.

    From => $admin_address
      "From" holds an e-mail address where the client's administrator and/or
      maintainer may be reached. It defaults to undef, which means no From
      header will be included in requests.

    MaxSize => OCTETS
      "MaxSize" specifies the largest response to accept from a server. The
      content of larger responses will be truncated to OCTET octets. This
      has been used to return the <head></head> section of web pages without
      the need to wade through <body></body>.

    NoProxy => [ $host_1, $host_2, ..., $host_N ]
    NoProxy => "host1,host2,hostN"
      "NoProxy" specifies a list of server hosts that will not be proxied.
      It is useful for local hosts and hosts that do not properly support
      proxying. If NoProxy is not specified, a list will be taken from the
      NO_PROXY environment variable.

        NoProxy => [ "localhost", "127.0.0.1" ],
        NoProxy => "localhost,127.0.0.1",

    BindAddr => $local_ip
      Specify "BindAddr" to bind all client sockets to a particular local
      address. The value of BindAddr will be passed through
      POE::Component::Client::Keepalive to POE::Wheel::SocketFactory (as
      "bind_address"). See that module's documentation for implementation
      details.

        BindAddr => "12.34.56.78"

    Protocol => $http_protocol_string
      "Protocol" advertises the protocol that the client wishes to see.
      Under normal circumstances, it should be left to its default value:
      "HTTP/1.1".

    Proxy => [ $proxy_host, $proxy_port ]
    Proxy => $proxy_url
    Proxy => $proxy_url,$proxy_url,...
      "Proxy" specifies one or more proxy hosts that requests will be passed
      through. If not specified, proxy servers will be taken from the
      HTTP_PROXY (or http_proxy) environment variable. No proxying will
      occur unless Proxy is set or one of the environment variables exists.

      The proxy can be specified either as a host and port, or as one or
      more URLs. Proxy URLs must specify the proxy port, even if it is 80.

        Proxy => [ "127.0.0.1", 80 ],
        Proxy => "http://127.0.0.1:80/",

      "Proxy" may specify multiple proxies separated by commas.
      PoCo::Client::HTTP will choose proxies from this list at random. This
      is useful for load balancing requests through multiple gateways.

        Proxy => "http://127.0.0.1:80/,http://127.0.0.1:81/",

    Streaming => OCTETS
      "Streaming" changes allows Client::HTTP to return large content in
      chunks (of OCTETS octets each) rather than combine the entire content
      into a single HTTP::Response object.

      By default, Client::HTTP reads the entire content for a response into
      memory before returning an HTTP::Response object. This is obviously
      bad for applications like streaming MP3 clients, because they often
      fetch songs that never end. Yes, they go on and on, my friend.

      When "Streaming" is set to nonzero, however, the response handler
      receives chunks of up to OCTETS octets apiece. The response handler
      accepts slightly different parameters in this case. ARG0 is also an
      HTTP::Response object but it does not contain response content, and
      ARG1 contains a a chunk of raw response content, or undef if the
      stream has ended.

        sub streaming_response_handler {
          my $response_packet = $_[ARG1];
          my ($response, $data) = @$response_packet;
          print SAVED_STREAM $data if defined $data;
        }

    FollowRedirects => $number_of_hops_to_follow
      "FollowRedirects" specifies how many redirects (e.g. 302 Moved) to
      follow. If not specified defaults to 0, and thus no redirection is
      followed. This maintains compatibility with the previous behavior,
      which was not to follow redirects at all.

      If redirects are followed, a response chain should be built, and can
      be accessed through $response_object->previous(). See HTTP::Response
      for details here.

    Timeout => $query_timeout
      "Timeout" sets how long POE::Component::Client::HTTP has to process an
      application's request, in seconds. "Timeout" defaults to 180 (three
      minutes) if not specified.

      It's important to note that the timeout begins when the component
      receives an application's request, not when it attempts to connect to
      the web server.

      Timeouts may result from sending the component too many requests at
      once. Each request would need to be received and tracked in order.
      Consider this:

        $_[KERNEL]->post(component => request => ...) for (1..15_000);

      15,000 requests are queued together in one enormous bolus. The
      component would receive and initialize them in order. The first socket
      activity wouldn't arrive until the 15,000th request was set up. If
      that took longer than "Timeout", then the requests that have waited
      too long would fail.

      "ConnectionManager"'s own timeout and concurrency limits also affect
      how many requests may be processed at once. For example, most of the
      15,000 requests would wait in the connection manager's pool until
      sockets become available. Meanwhile, the "Timeout" would be counting
      down.

      Applications may elect to control concurrency outside the component's
      "Timeout". They may do so in a few ways.

      The easiest way is to limit the initial number of requests to
      something more manageable. As responses arrive, the application should
      handle them and start new requests. This limits concurrency to the
      initial request count.

      An application may also outsource job throttling to another module,
      such as POE::Component::JobQueue.

      In any case, "Timeout" and "ConnectionManager" may be tuned to
      maximize timeouts and concurrency limits. This may help in some cases.
      Developers should be aware that doing so will increase memory usage.
      POE::Component::Client::HTTP and KeepAlive track requests in memory,
      while applications are free to keep pending requests on disk.

ACCEPTED EVENTS
    Sessions communicate asynchronously with PoCo::Client::HTTP. They post
    requests to it, and it posts responses back.

  request
    Requests are posted to the component's "request" state. They include an
    HTTP::Request object which defines the request. For example:

      $kernel->post(
        'ua', 'request',            # http session alias & state
        'response',                 # my state to receive responses
        GET('http://poe.perl.org'), # a simple HTTP request
        'unique id',                # a tag to identify the request
        'progress',                 # an event to indicate progress
        'http://1.2.3.4:80/'        # proxy to use for this request
      );

    Requests include the state to which responses will be posted. In the
    previous example, the handler for a 'response' state will be called with
    each HTTP response. The "progress" handler is optional and if installed,
    the component will provide progress metrics (see sample handler below).
    The "proxy" parameter is optional and if not defined, a default proxy
    will be used if configured. No proxy will be used if neither a default
    one nor a "proxy" parameter is defined.

  pending_requests_count
    There's also a pending_requests_count state that returns the number of
    requests currently being processed. To receive the return value, it must
    be invoked with $kernel->call().

      my $count = $kernel->call('ua' => 'pending_requests_count');

    NOTE: Sometimes the count might not be what you expected, because
    responses are currently in POE's queue and you haven't processed them.
    This could happen if you configure the "ConnectionManager"'s concurrency
    to a high enough value.

  cancel
    Cancel a specific HTTP request. Requires a reference to the original
    request (blessed or stringified) so it knows which one to cancel. See
    "progress handler" below for notes on canceling streaming requests.

    To cancel a request based on its blessed HTTP::Request object:

      $kernel->post( component => cancel => $http_request );

    To cancel a request based on its stringified HTTP::Request object:

      $kernel->post( component => cancel => "$http_request" );

  shutdown
    Responds to all pending requests with 408 (request timeout), and then
    shuts down the component and all subcomponents.

SENT EVENTS
  response handler
    In addition to all the usual POE parameters, HTTP responses come with
    two list references:

      my ($request_packet, $response_packet) = @_[ARG0, ARG1];

    $request_packet contains a reference to the original HTTP::Request
    object. This is useful for matching responses back to the requests that
    generated them.

      my $http_request_object = $request_packet->[0];
      my $http_request_tag    = $request_packet->[1]; # from the 'request' post

    $response_packet contains a reference to the resulting HTTP::Response
    object.

      my $http_response_object = $response_packet->[0];

    Please see the HTTP::Request and HTTP::Response manpages for more
    information.

  progress handler
    The example progress handler shows how to calculate a percentage of
    download completion.

      sub progress_handler {
        my $gen_args  = $_[ARG0];    # args passed to all calls
        my $call_args = $_[ARG1];    # args specific to the call

        my $req = $gen_args->[0];    # HTTP::Request object being serviced
        my $tag = $gen_args->[1];    # Request ID tag from.
        my $got = $call_args->[0];   # Number of bytes retrieved so far.
        my $tot = $call_args->[1];   # Total bytes to be retrieved.
        my $oct = $call_args->[2];   # Chunk of raw octets received this time.

        my $percent = $got / $tot * 100;

        printf(
          "-- %.0f%% [%d/%d]: %s\n", $percent, $got, $tot, $req->uri()
        );

        # To cancel the request:
        # $_[KERNEL]->post( component => cancel => $req );
      }

   DEPRECATION WARNING
    The third return argument (the raw octets received) has been deprecated.
    Instead of it, use the Streaming parameter to get chunks of content in
    the response handler.

REQUEST CALLBACKS
    The HTTP::Request object passed to the request event can contain a CODE
    reference as "content". This allows for sending large files without
    wasting memory. Your callback should return a chunk of data each time it
    is called, and an empty string when done. Don't forget to set the
    Content-Length header correctly. Example:

      my $request = HTTP::Request->new( PUT => 'http://...' );

      my $file = '/path/to/large_file';

      open my $fh, '<', $file;

      my $upload_cb = sub {
        if ( sysread $fh, my $buf, 4096 ) {
          return $buf;
        }
        else {
          close $fh;
          return '';
        }
      };

      $request->content_length( -s $file );

      $request->content( $upload_cb );

      $kernel->post( ua => request, 'response', $request );

CONTENT ENCODING AND COMPRESSION
    Transparent content decoding has been disabled as of version 0.84. This
    also removes support for transparent gzip requesting and decompression.

    To re-enable gzip compression, specify the gzip Content-Encoding and use
    HTTP::Response's decoded_content() method rather than content():

      my $request = HTTP::Request->new(
        GET => "http://www.yahoo.com/", [
          'Accept-Encoding' => 'gzip'
        ]
      );

      # ... time passes ...

      my $content = $response->decoded_content();

    The change in POE::Component::Client::HTTP behavior was prompted by
    changes in HTTP::Response that surfaced a bug in the component's
    transparent gzip handling.

    Allowing the application to specify and handle content encodings seems
    to be the most reliable and flexible resolution.

    For more information about the problem and discussions regarding the
    solution, see: <http://www.perlmonks.org/?node_id=683833> and
    <http://rt.cpan.org/Ticket/Display.html?id=35538>

CLIENT HEADERS
    POE::Component::Client::HTTP sets its own response headers with
    additional information. All of its headers begin with "X-PCCH".

  X-PCCH-Errmsg
    POE::Component::Client::HTTP may fail because of an internal client
    error rather than an HTTP protocol error. X-PCCH-Errmsg will contain a
    human readable reason for client failures, should they occur.

    The text of X-PCCH-Errmsg may also be repeated in the response's
    content.

  X-PCCH-Peer
    X-PCCH-Peer contains the remote IPv4 address and port, separated by a
    period. For example, "127.0.0.1.8675" represents port 8675 on localhost.

    Proxying will render X-PCCH-Peer nearly useless, since the socket will
    be connected to a proxy rather than the server itself.

    This feature was added at Doreen Grey's request. Doreen wanted a means
    to find the remote server's address without having to make an additional
    request.

ENVIRONMENT
    POE::Component::Client::HTTP uses two standard environment variables:
    HTTP_PROXY and NO_PROXY.

    HTTP_PROXY sets the proxy server that Client::HTTP will forward requests
    through. NO_PROXY sets a list of hosts that will not be forwarded
    through a proxy.

    See the Proxy and NoProxy constructor parameters for more information
    about these variables.

SEE ALSO
    This component is built upon HTTP::Request, HTTP::Response, and POE.
    Please see its source code and the documentation for its foundation

CHANGES
==================================================
Changes from 2012-05-06 00:00:00 +0000 to present.
==================================================

------------------------------------------
version 0.948 at 2013-05-06 17:22:31 +0000
------------------------------------------

  Change: c60d4d55461318cf20937748bae218db00f53418
  Author: Rocco Caputo <rcaputo@cpan.org>
  Date : 2013-05-06 13:17:29 +0000

    Return a "400 Bad Request" error if we're redirected to a URI with a
    bad scheme.

    Sean Brunnock reported the issue when a link redirected his program
    to <htttp://...>. This triggered an internal confess() we had put
    there to trap a strange case we couldn't reproduce. His test case and
    patch were instrumental in helping me resolve this on a deeper level. 

  Change: ced552540af225632756d5696aafee17e0876cf8
  Author: Theron Stanford <shixilun@gmail.com>
  Date : 2013-01-13 15:19:55 +0000

    Fix a doc type; (GET x, y) is not (GET(x), y). 

------------------------------------------
version 0.947 at 2012-06-03 07:21:46 +0000
------------------------------------------

  Change: 944c66f2149bc39a24db854d0310522b1d1697ed
  Author: Rocco Caputo <rcaputo@cpan.org>
  Date : 2012-06-03 03:21:46 +0000

    [rt.cpan.org 76542] Fix a double close in PoCo::Client::HTTP.

    Florian Schlichting found POE throwing an error on an undefined file
    descriptor. POE::Component::Client::HTTP was closing a filehandle
    before destroying the wheel that was managing it. When it was time
    for the wheel to remove the filehandle's watcher, fileno() failed,
    and trouble ensued. 

  Change: 7628a0aa4c7097a54d46773d3bdc3000362a0fd1
  Author: Rocco Caputo <rcaputo@cpan.org>
  Date : 2012-06-03 02:27:03 +0000

    [rt.cpan.org 77601] Switch to Socket's getaddrinfo().

    AVENJ reported that Socket::GetSockAddr's :newapi was deprecated.
    Switch to Socket for a more future-proof API. 

------------------------------------------
version 0.946 at 2012-05-15 01:47:32 +0000
------------------------------------------

  Change: eacdcf93f62f566011a8c29637d10f27d0aad28a
  Author: Rocco Caputo <rcaputo@cpan.org>
  Date : 2012-05-14 21:47:32 +0000

    Rely upon the latest POE::Component::Client::Keepalive. 

  Change: 567c9bf1b2a7ac5aa057baaefa545055548dafc2
  Author: Rocco Caputo <rcaputo@cpan.org>
  Date : 2012-04-29 19:50:38 +0000

    [rt.cpan.org 76543] Help out Debian with a NAME and SYNOPSIS for an
    internal module.

    Florian sent in a patch written by Krzysztof Krzyzaniak to add NAME
    and SYNOPSIS sections to
    POE::Component::Client::HTTP::RequestFactory. This is reported to
    help Debian automatically build a "whatis" entry for the module.
    Thanks for the patch! Please send any others upstream. 

=================================================
Plus 58 releases after 2012-05-06 00:00:00 +0000.
=================================================


Request.pm
# $Id: Request.pm 203 2005-08-02 16:02:33Z rcaputo $

package POE::Component::Client::HTTP::Request;
use strict;
use warnings;

use POE;

use Carp;
use HTTP::Status;

# Unique request ID, independent of wheel and timer IDs.
my $request_seq = 0;

sub DEBUG () { 0 }

sub REQ_ID            () {  0 }
sub REQ_POSTBACK      () {  1 }
sub REQ_CONNECTION    () {  2 }
sub REQ_REQUEST       () {  3 }
sub REQ_STATE         () {  4 }
sub REQ_RESPONSE      () {  5 }
sub REQ_BUFFER        () {  6 }

sub REQ_OCTETS_GOT    () {  8 }
sub REQ_TIMER         () {  9 }
sub REQ_PROG_POSTBACK () { 10 }
sub REQ_USING_PROXY   () { 11 }
sub REQ_HOST          () { 12 }
sub REQ_PORT          () { 13 }
sub REQ_HISTORY       () { 14 }
sub REQ_START_TIME    () { 15 }
sub REQ_FACTORY       () { 16 }

sub RS_CONNECT      () { 0x01 }
sub RS_SENDING      () { 0x02 }
sub RS_IN_HEAD      () { 0x04 }
sub RS_REDIRECTED   () { 0x08 }
sub RS_IN_CONTENT   () { 0x20 }
sub RS_DONE         () { 0x40 }
sub RS_POSTED       () { 0x80 }

sub import {
  my ($class) = shift;

  my $package = caller();

  foreach my $tag (@_) {
    if ($tag eq ':fields') {
      foreach my $sub (
        qw(
          REQ_ID REQ_POSTBACK REQ_CONNECTION REQ_REQUEST REQ_STATE
          REQ_RESPONSE REQ_BUFFER REQ_OCTETS_GOT REQ_TIMER
          REQ_PROG_POSTBACK REQ_USING_PROXY REQ_HOST REQ_PORT
          REQ_HISTORY REQ_START_TIME
        )
      ) {
        no strict 'refs';
        *{$package . "::$sub"} = \&$sub;
      }
    }

    if ($tag eq ':states') {
      foreach my $sub (
        qw(
          RS_CONNECT RS_SENDING RS_IN_HEAD RS_REDIRECTED
          RS_IN_CONTENT RS_DONE RS_POSTED
        )
      ) {
        no strict 'refs';
        *{$package . "::$sub"} = \&$sub;
      }
    }
  }
}

sub ID {
  my ($self) = @_;
  return $self->[REQ_ID];
}

sub new {
  my $class = shift;

  croak __PACKAGE__ . "expects its arguments to be key/value pairs" if @_ & 1;
  my %params = @_;

  croak "need a Request parameter" unless (defined $params{'Request'});
  croak "Request must be a HTTP::Request object"
    unless (UNIVERSAL::isa ($params{'Request'}, "HTTP::Request"));

  #croak "need a Tag parameter" unless (defined $params{'Tag'});

  croak "need a Factory parameter" unless (defined $params{'Factory'});

  my ($http_request, $postback, $tag, $progress, $factory) =
    @params{qw(Request Postback Tag Progress Factory)};

  my $request_id = ++$request_seq;
  DEBUG and warn "REQ: creating a request ($request_id)";

  # Get the host and port from the request object.
  my ($host, $port, $scheme, $using_proxy);

  eval {
    $host   = $http_request->uri()->host();
    $port   = $http_request->uri()->port();
    $scheme = $http_request->uri()->scheme();
  };
  croak "Not a usable Request: $@" if ($@);

  # Add a host header if one isn't included.  Must do this before
  # we reset the $host for the proxy!
  _set_host_header ($http_request) unless (
    defined $http_request->header('Host') and
    length $http_request->header('Host')
  );


  if (defined $params{Proxy}) {
    # This request qualifies for proxying.  Replace the host and port
    # with the proxy's host and port.  This comes after the Host:
    # header is set, so it doesn't break the request object.
    ($host, $port) = @{$params{Proxy}->[rand @{$params{Proxy}}]};

    $using_proxy = 1;
  }
  else {
    $using_proxy = 0;
  }

  # Build the request.
  my $self = [
    $request_id,        # REQ_ID
    $postback,          # REQ_POSTBACK
    undef,              # REQ_CONNECTION
    $http_request,      # REQ_REQUEST
    RS_CONNECT,         # REQ_STATE
    undef,              # REQ_RESPONSE
    '',                 # REQ_BUFFER
    undef,              # unused
    0,                  # REQ_OCTETS_GOT
    undef,              # REQ_TIMER
    $progress,          # REQ_PROG_POSTBACK
    $using_proxy,       # REQ_USING_PROXY
    $host,              # REQ_HOST
    $port,              # REQ_PORT
    undef,              # REQ_HISTORY
    time(),             # REQ_START_TIME
    $factory,           # REQ_FACTORY
   ];
   return bless $self, $class;
}

sub return_response {
  my ($self) = @_;

  DEBUG and warn "in return_response ", sprintf ("0x%02X", $self->[REQ_STATE]);
  return if ($self->[REQ_STATE] & RS_POSTED);
  my $response = $self->[REQ_RESPONSE];

  # If we have a cookie jar, have it frob our headers.  LWP rocks!
  $self->[REQ_FACTORY]->frob_cookies ($response);

  # If we're done, send back the HTTP::Response object, which
  # is filled with content if we aren't streaming, or empty
  # if we are. that there's no ARG1 lets the client know we're done
  # with the content in the latter case
  if ($self->[REQ_STATE] & RS_DONE) {
    DEBUG and warn "done; returning $response for ", $self->[REQ_ID];
    $self->[REQ_POSTBACK]->($self->[REQ_RESPONSE]);
    $self->[REQ_STATE] |= RS_POSTED;
    #warn "state is now ", $self->[REQ_STATE];
  }
  elsif ($self->[REQ_STATE] & RS_IN_CONTENT) {
    # If we are streaming, send the chunk back to the client session.
    # Otherwise add the new octets to the response's content.
    # This should only add up to content-length octets total!
    if ($self->[REQ_FACTORY]->is_streaming) {
      DEBUG and warn "returning partial $response";
      $self->[REQ_POSTBACK]->($self->[REQ_RESPONSE], $self->[REQ_BUFFER]);
    }
    else {
      DEBUG and warn "adding to $response";
      $self->[REQ_RESPONSE]->add_content($self->[REQ_BUFFER]);
    }
  }
  $self->[REQ_BUFFER] = '';
}

sub add_eof {
  my ($self) = @_;

  return if ($self->[REQ_STATE] & RS_POSTED);

  unless (defined $self->[REQ_RESPONSE]) {
    # XXX I don't know if this is actually used
    $self->error(400, "incomplete response " . $self->[REQ_ID]);
    return;
  }

  # RFC 2616: "If a message is received with both a Transfer-Encoding
  # header field and a Content-Length header field, the latter MUST be
  # ignored."
  if (
    defined $self->[REQ_RESPONSE]->content_length and
    not defined $self->[REQ_RESPONSE]->header("Transfer-Encoding") and
    $self->[REQ_OCTETS_GOT] < $self->[REQ_RESPONSE]->content_length
  ) {
    DEBUG and warn(
      "got " . $self->[REQ_OCTETS_GOT] . " of " .
      $self->[REQ_RESPONSE]->content_length
    );
    $self->error(400, "incomplete response " . $self->[REQ_ID]);
  }
  else {
    $self->[REQ_STATE] |= RS_DONE;
    $self->return_response();
  }
}

sub add_content {
  my ($self, $data) = @_;

  if (ref $data) {
    $self->[REQ_STATE] = RS_DONE;
    $data->scan (sub {$self->[REQ_RESPONSE]->header (@_) });
    return 1;
  }

  $self->[REQ_BUFFER] .= $data;

  # Count how many octets we've received.  -><- This may fail on
  # perl 5.8 if the input has been identified as Unicode.  Then
  # again, the C<use bytes> in Driver::SysRW may have untainted the
  # data... or it may have just changed the semantics of length()
  # therein.  If it's done the former, then we're safe.  Otherwise
  # we also need to C<use bytes>.
  # TODO: write test(s) for this.
  my $this_chunk_length = length($self->[REQ_BUFFER]);
  $self->[REQ_OCTETS_GOT] += $this_chunk_length;

  my $max = $self->[REQ_FACTORY]->max_response_size;

  DEBUG and warn(
    "REQ: request ", $self->ID,
    " received $self->[REQ_OCTETS_GOT] bytes; maximum is $max"
  );

  if (defined $max and $self->[REQ_OCTETS_GOT] > $max) {
    # We've gone over the maximum content size to return.  Chop it # back.
    my $over = $self->[REQ_OCTETS_GOT] - $max;
    $self->[REQ_OCTETS_GOT] -= $over;
    substr($self->[REQ_BUFFER], -$over) = "";
    #TODO: ??
    #$self->[REQ_STATE] |= RS_DONE;
  }

  # keep this for the progress callback (it gets cleared in return_response
  # as I say below, this needs to go away.
  my $buffer = $self->[REQ_BUFFER];

  $self->return_response;
  DEBUG and do {
    warn(
      "REQ: request ", $self->ID,
      " got $this_chunk_length octets of content..."
    );

    warn(
      "REQ: request ", $self->ID, " has $self->[REQ_OCTETS_GOT]",
      (
        $self->[REQ_RESPONSE]->content_length()
        ? ( " out of " . $self->[REQ_RESPONSE]->content_length() )
        : ""
      ),
      " octets"
    );
  };

  if (defined $max and $self->[REQ_OCTETS_GOT] >= $max) {
    DEBUG and warn(
      "REQ: request ", $self->ID, " has a full response... moving to done."
    );
    $self->[REQ_STATE] |= RS_DONE;
    $self->[REQ_STATE] &= ~RS_IN_CONTENT;
    return 1;
  }

  if ($self->[REQ_RESPONSE]->content_length) {

    # Report back progress
    $self->[REQ_PROG_POSTBACK]->(
      $self->[REQ_OCTETS_GOT],
      $self->[REQ_RESPONSE]->content_length,
      #TODO: ugh. this is stupid. Must remove/deprecate!
      $buffer,
    ) if ($self->[REQ_PROG_POSTBACK]);

    # Stop reading when we have enough content.  -><- Should never be
    # greater than our content length.
    if ($self->[REQ_OCTETS_GOT] >= $self->[REQ_RESPONSE]->content_length) {
      DEBUG and warn(
        "REQ: request ", $self->ID, " has a full response... moving to done."
      );
      $self->[REQ_STATE] |= RS_DONE;
      $self->[REQ_STATE] &= ~RS_IN_CONTENT;
      return 1;
    }
  }
  return 0;
}

sub timer {
  my ($self, $timer) = @_;

  # do it this way so we can set REQ_TIMER to undef
  if (@_ == 2) {
    die "overwriting timer $self->[REQ_TIMER]" if $self->[REQ_TIMER];
    $self->[REQ_TIMER] = $timer;
  }
  return $self->[REQ_TIMER];
}

sub create_timer {
  my ($self, $timeout) = @_;

  # remove old timeout first
  my $kernel = $POE::Kernel::poe_kernel;

  my $seconds = $timeout - (time() - $self->[REQ_START_TIME]);
  $self->[REQ_TIMER] = $kernel->delay_set(
    got_timeout => $seconds, $self->ID
  );
  DEBUG and warn(
    "TKO: request ", $self->ID,
    " has timer $self->[REQ_TIMER] going off in $seconds seconds\n"
  );
}

sub remove_timeout {
  my ($self) = @_;

  my $alarm_id = $self->[REQ_TIMER];
  if (defined $alarm_id) {
    my $kernel = $POE::Kernel::poe_kernel;
    DEBUG and warn "REQ: Removing timer $alarm_id";
    $kernel->alarm_remove($alarm_id);
    $self->[REQ_TIMER] = undef;
  }
}

sub postback {
  my ($self, $postback) = @_;

  if (defined $postback) {
    DEBUG and warn "REQ: modifying postback";
    $self->[REQ_POSTBACK] = $postback;
  }
  return $self->[REQ_POSTBACK];
}

sub _set_host_header {
  my ($request) = @_;
  my $uri = $request->uri;

  my ($new_host, $new_port);
  eval {
    $new_host = $uri->host();
    $new_port = $uri->port();
    # Only include the port if it's nonstandard.
    if ($new_port == 80 || $new_port == 443) {
      $request->header( Host => $new_host );
    }
    else {
      $request->header( Host => "$new_host:$new_port" );
    }
  };
  warn $@ if $@;
}

sub does_redirect {
  my ($self, $last) = @_;

  if (defined $last) {
    $self->[REQ_HISTORY] = $last;
    # delete OLD timeout
    #my $alarm_id = $last->[REQ_TIMEOUT];
    #DEBUG and warn "RED: Removing old timeout $alarm_id\n";
    #$POE::Kernel::poe_kernel->alarm_remove ($alarm_id);
  }

  return defined $self->[REQ_HISTORY];
}

sub check_redirect {
  my ($self) = @_;

  my $max = $self->[REQ_FACTORY]->max_redirect_count;

  if (defined $self->[REQ_HISTORY]) {
    $self->[REQ_RESPONSE]->previous($self->[REQ_HISTORY]->[REQ_RESPONSE]);
  }

  return undef unless ($self->[REQ_RESPONSE]->is_redirect);

  my $new_uri = $self->[REQ_RESPONSE]->header ('Location');
  DEBUG and warn "REQ: Preparing redirect to $new_uri";
  my $base = $self->[REQ_RESPONSE]->base();
  $new_uri = URI->new($new_uri, $base)->abs($base);
  DEBUG and warn "RED: Actual redirect uri is $new_uri";

  my $prev = $self;
  my $history = 0;
  while ($prev = $prev->[REQ_HISTORY]) {
    $history++;
    $history = $max + 1 if ($prev->[REQ_REQUEST]->uri eq $new_uri);
    last if ($history > $max);
  }

  if ($history > $max) {
    #$self->[REQ_STATE] |= RS_DONE;
    DEBUG and warn "RED: Too much redirection";
  }
  else { # All fine, yield new request and mark this disabled.
    my $newrequest = $self->[REQ_REQUEST]->clone();
    DEBUG and warn "RED: new request $newrequest";
    $newrequest->uri($new_uri);
    _set_host_header ($newrequest);

    $self->[REQ_STATE] = RS_REDIRECTED;
    DEBUG and warn "RED: new request $newrequest";
    return $newrequest;
  }
  return undef;
}

sub send_to_wheel {
  my ($self) = @_;

  $self->[REQ_STATE] = RS_SENDING;

  my $http_request = $self->[REQ_REQUEST];

  # MEXNIX 2002-06-01: Check for proxy.  Request query is a bit
  # different...

  my $request_uri;
  if ($self->[REQ_USING_PROXY]) {
    $request_uri = $http_request->uri()->canonical();
  }
  else {
    $request_uri = $http_request->uri()->canonical()->path_query();
  }

  my $request_string = (
    $http_request->method() . ' ' .
    $request_uri . ' ' .
    $http_request->protocol() . "\x0D\x0A" .
    $http_request->headers_as_string("\x0D\x0A") . "\x0D\x0A" .
    $http_request->content() # . "\x0D\x0A"
  );

  DEBUG and do {
    my $formatted_request_string = $request_string;
    $formatted_request_string =~ s/([^\n])$/$1\n/;
    $formatted_request_string =~ s/^/| /mg;
    warn ",----- SENDING REQUEST ", '-' x 56, "\n";
    warn $formatted_request_string;
    warn "`", '-' x 78, "\n";
  };

  $self->[REQ_CONNECTION]->wheel->put ($request_string);
}

sub wheel {
  my ($self) = @_;

  #if (defined $new_wheel) {
    # Switch wheels.  This is cumbersome, but it works around a bug in
    # older versions of POE.
    #$self->[REQ_WHEEL] = undef;
    #$self->[REQ_WHEEL] = $new_wheel;
  #}
  return $self->[REQ_CONNECTION]->wheel;
}

sub error {
  my ($self, $code, $message) = @_;

  my $nl = "\n";

  my $r = HTTP::Response->new($code);
  my $http_msg = status_message ($code);
  my $m = (
    "<html>$nl"
    . "<HEAD><TITLE>Error: $http_msg</TITLE></HEAD>$nl"
    . "<BODY>$nl"
    . "<H1>Error: $http_msg</H1>$nl"
    . "$message$nl"
    . "</BODY>$nl"
    . "</HTML>$nl"
  );

  $r->content ($m);
  $self->[REQ_POSTBACK]->($r);
}

sub connect_error {
  my ($self, $message) = @_;

  my $host = $self->[REQ_HOST];
  my $port = $self->[REQ_PORT];

  $message = "Cannon connect to $host:$port ($message)";
  $self->error (RC_INTERNAL_SERVER_ERROR, $message);
  return;
}

sub host { shift->[REQ_HOST] }

sub port { shift->[REQ_PORT] }

sub scheme {
  my $self = shift;

  $self->[REQ_USING_PROXY] ? 'http' : $self->[REQ_REQUEST]->uri->scheme;
}

sub DESTROY {
  my ($self) = @_;

  delete $self->[REQ_CONNECTION];
  delete $self->[REQ_FACTORY];
}

1;
